app/controllers/users_controller.rb

```diff
@@ -290,8 +241,23 @@ def auth_success
         else
           failed_login t("sessions.new.auth failure")
         end
+      elsif user.nil? && user = User.find_by(:email => email)
+        user[:auth_uid] = uid
+        user[:auth_provider] = provider
+        user.save!
+
+        user.deactivate! if user.status == "active" && !email_verified
+
+        if user.status == "active"
+          successful_login(user)
+        else
+          session[:token] = user.tokens.create.token
+          UserMailer.signup_confirm(user, user.tokens.create(:referer => session[:referer])).deliver_later
+          redirect_to :controller => :confirmations, :action => :confirm, :display_name => user.display_name
+        end
```

Comment:
This is trying to deal with the case where somebody does a social signup and there is already an account with the email address the social provider returns and it looks problematic to me for a number of reasons:

* If the user already had a different social provider attached to their account this will overwrite it
* If the social provider satisfies our (deliberately weak) definition of providing a verified email then we allow them to attach to the existing account with no further verification but that is very different to allowing creation a new account based on a weak verification
* If the social provider email is not considered verified then we send a normal signup confirmation but the working of that will be a bit odd in this case I think plus we actual deactivate the account until that confirmation happens

There's a definite problem here, that I had a case of only this week, where somebody tries a social signup and they already have an account but I think we need to send a confirmation email in all cases and probably make it a different one saying that we need to confirm the social signup link and just use it to confirm the link rather than doing a deactivate first.

I'm not sure what to do where there is already a different provider attached though though arguably that's fine if we only change the auth provider after email confirmation and the email says we will be replacing the old one.

Reply:
I agree with you that this is indeed a tricky scenario. I am not really a fan of hijacking the existing account once you signin successfully with social account with same email address.

> There's a definite problem here, that I had a case of only this week, where somebody tries a social signup and they already have an account but I think we need to send a confirmation email in all cases and probably make it a different one saying that we need to confirm the social signup link and just use it to confirm the link rather than doing a deactivate first.

I suspect user already had an OSM account and tried logging in with social account, with desire to add an alternate method of signing in. The functionality to do this already exists, by modifying in the "External Authentication" in settings.

> There's a definite problem here, that I had a case of only this week, where somebody tries a social signup and they already have an account but I think we need to send a confirmation email in all cases and probably make it a different one saying that we need to confirm the social signup link and just use it to confirm the link rather than doing a deactivate first.

This seems like a complex approach to me, since the confirmation email process currently practically only activates the account.

I suggest we step back and simply detect this scenario, and suggest the user to actually login with their email account/password, and modify their settings to include social sign in? This would also eliminate the scenarios where the user already uses a different social signup.

Reply:
Yes that is certainly the simplest solution, and removes the problems with the current code.

Reply:
I have reverted the "hijacking of the OSM account by social sign-in" feature. Instead, I only detect the condition and display a message as the flash notice. Currently the text says:

> If you already have OpenStreetMap account and wish to use 3rd party identity provider, please login using your password and modify settings of your account.